decryptor/postgresql/prepared_statements.go

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2020, Cossack Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package postgresql
+
+import (
+	"errors"
+
+	"github.com/cossacklabs/acra/decryptor/base"
+	"github.com/cossacklabs/acra/sqlparser"
+)
+
+// Errors returned by prepared statement registry.
+var (
+	ErrStatementNotFound = errors.New("no prepared statement with given name")
+	ErrCursorNotFound    = errors.New("no cursor with given name")
+)
+
+// PgPreparedStatementRegistry is a PostgreSQL PreparedStatementRegistry.
+type PgPreparedStatementRegistry struct {
+	registry map[string]base.PreparedStatement
+}
+
+// NewPreparedStatementRegistry makes a new empty prepared statement registry.
+func NewPreparedStatementRegistry() *PgPreparedStatementRegistry {
+	return &PgPreparedStatementRegistry{registry: make(map[string]base.PreparedStatement)}
+}
+
+// StatementByName returns a prepared statement from the registry by its name, if it exists.
+func (r *PgPreparedStatementRegistry) StatementByName(name string) (base.PreparedStatement, error) {
+	s, ok := r.registry[name]
+	if ok {
+		return s, nil
+	}
+	return nil, ErrStatementNotFound
+}
+
+// Add a statement to the registry. If an existing statement with the same name exists,
+// it is replaced with the new one. Returns "true" if statement has been replaced.
+func (r *PgPreparedStatementRegistry) Add(statement base.PreparedStatement) (bool, error) {
+	name := statement.Name()
+	_, exists := r.registry[name]
+	r.registry[name] = statement
+	return exists, nil
+}
+
+// PgPreparedStatement is a PostgreSQL PreparedStatement.
+type PgPreparedStatement struct {
+	name string
+	text string
+	sql  sqlparser.Statement
```

Comment:
just to think about it. probably we will need to add some context storage for statement. for example bound values or our encryption settings related to this prepared statement (but probably not if we will do it in another way and search encryption settings by prepared statement name in other registry)